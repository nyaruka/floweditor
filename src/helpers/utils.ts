import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';
import { CharacterSet } from '../components/form/TextInputElement';

const SNAKED_CHARS = /\s+(?=\S)/g;
const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export const getDisplayName = (WrappedComponent: ComponentClass | SFC): string =>
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 * @param {Array.<string>} errors - list of error messages
 * @param {Array.<string>} queries  - list of queries
 * @returns {boolean}
 */
export const hasErrorType = (errors: string[], queries: string[]): boolean => {
    if (!errors.length) {
        return false;
    }

    for (const error of errors) {
        for (const query of queries) {
            if (error.indexOf(query) > -1) {
                return true;
            }
        }
    }

    return false;
};
