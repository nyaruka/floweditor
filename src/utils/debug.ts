import { DebugState } from 'store/editor';
import { store } from 'store';

const mutate = require('immutability-helper');

/* istanbul ignore next */
export default class Debug {
  private state: DebugState;

  constructor(props: any, initial: DebugState) {
    this.state = initial || { showUUIDs: false };
  }

  public showUUIDs(): DebugState {
    const updated = mutate(this.state, { $merge: { showUUIDs: true } });
    
    // Update via TembaStore instead of Redux
    const currentEditorState = store.getState().editorState;
    store.getState().updateEditorState({ 
      ...currentEditorState, 
      debug: updated 
    });
    
    return updated;
  }
}
