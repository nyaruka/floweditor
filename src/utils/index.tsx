import * as React from 'react';
import { Language } from '../component/LanguageSelector';
import { Action, Case, Exit, Languages, LocalizationMap, ContactProperties } from '../flowTypes';
import Localization, { LocalizedObject } from '../services/Localization';
import { SearchResult } from '../store';
import * as variables from '../variables.scss';

export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
const LABEL_CHARS = /^[a-zA-Z0-9-][a-zA-Z0-9- ]*$/;
const MAX_LABEL_LENGTH = 36;
const SNAKED_CHARS = /\s+(?=\S)/g;

export const GRID_SIZE: number = parseInt(variables.grid_size, 10) || 20;
export const NODE_SPACING: number = parseInt(variables.node_spacing, 10) || 10;
export const NODE_PADDING: number = parseInt(variables.node_padding, 10) || 10;
export const QUIET_NOTE = 2000;
export const CONFIRMATION_TIME = 2000;

interface BoolMap {
    [key: string]: boolean;
}

interface Bounds {
    left: number;
    top: number;
    right: number;
    bottom: number;
}

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
    let leftAdjust = left % GRID_SIZE;
    let topAdjust = top % GRID_SIZE;

    if (leftAdjust > GRID_SIZE / 3) {
        leftAdjust = GRID_SIZE - leftAdjust;
    } else {
        leftAdjust = leftAdjust * -1;
    }

    if (topAdjust > GRID_SIZE / 3) {
        topAdjust = GRID_SIZE - topAdjust;
    } else {
        topAdjust = topAdjust * -1;
    }

    return {
        left: Math.max(left + leftAdjust, 0),
        top: Math.max(top + topAdjust, 0)
    };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');
/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
    if (!errors.length) {
        return false;
    }
    for (const error of errors) {
        for (const exp of exps) {
            if (error.match(exp)) {
                return true;
            }
        }
    }
    return false;
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
    if (localizations && localizations.length) {
        return localizations[0];
    }
};

export interface ClickHandler {
    onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void;
    onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Creates a simple click handler via onMouseDown and onMouseUp.
 * This is a necessity in order to let jsPlumb manage our element dragging.
 * @param onClick
 */
export const createClickHandler = (
    onClick: (event: React.MouseEvent<HTMLDivElement>) => void
): ClickHandler => {
    return {
        onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => {
            this._clicked = true;
        },
        onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => {
            if (this._clicked) {
                onClick(event);
            }
            this._clicked = false;
        }
    };
};

/**
 * Callback to pass to Array.prototype.map to return SearchResult[].
 * Use on 'results' property of payload returned from an endpoint returning
 * groups, fields.
 */
export const resultsToSearchOpts = ({ name, uuid, type }: any): SearchResult => ({
    name,
    id: uuid,
    type
});

/**
 * Get the first language in a Languages map
 */
export const getBaseLanguage = (languages: Languages): Language => {
    const [iso] = Object.keys(languages);
    const name = languages[iso];
    return {
        name,
        iso
    };
};

/**
 * Get a language from a Languages map in Language format
 */
export const getLanguage = (languages: Languages, iso: string): Language => ({
    name: languages[iso],
    iso
});

export const getLocalization = (
    obj: Action | Exit | Case,
    localization: LocalizationMap,
    iso: string,
    languages: Languages
) => Localization.translate(obj, iso, languages, localization[iso]);

export const dump = (thing: any) => console.log(JSON.stringify(thing, null, 2));

/**
 * Apply emphasis style
 */
export const emphasize = (text: string) => <span className="emph">{text}</span>;

/**
 * Does property exist in our ContactProperties enum?
 * @param {string} propertyToCheck - 'name' key on react-select AutoCompleteOption
 */
export const propertyExists = (propertyToCheck: string) => {
    for (const property of Object.keys(ContactProperties)) {
        if (property.toLowerCase() === propertyToCheck.toLowerCase().trim()) {
            return true;
        }
    }
    return false;
};

/**
 * Should x element be rendered?
 */
export const renderIf = (condition: boolean) => (elem: JSX.Element) => (condition ? elem : null);

/**
 * Does the label meet our length requirements?
 * @param {string} label - label created by react-select
 */
export const properLabelLength = (label: string = '') =>
    label.length > 0 && label.length <= MAX_LABEL_LENGTH;

/**
 * Does the label meet our character requirements?
 * @param {string} label
 */
export const containsOnlyLabelChars = (label: string = '') => LABEL_CHARS.test(label);

/**
 * Does the label meet our length, character requirements?
 * @param {string} label - label created by react-select
 */
export const isValidLabel = (label: string) =>
    properLabelLength(label) && containsOnlyLabelChars(label);
