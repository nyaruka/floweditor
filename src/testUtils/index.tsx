/* eslint-disable @typescript-eslint/no-explicit-any */
// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowTypes } from 'config/interfaces';
import { info } from 'core-js/core/log';
import { keyFor } from 'core-js/fn/symbol';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { loadStore } from 'store';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import { App, FlowContents, FlowInfo, InfoResult, TembaAppState } from 'temba-components';
import { getFlowEditorConfig } from 'test/config';
import * as matchers from 'testUtils/matchers';
import { createUUID, merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: Array<{ [key: string]: any }>;
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' }
  })
});

export const getConfigProviderContext = (
  flowType: FlowTypes = FlowTypes.MESSAGING
): ConfigProviderContext => {
  const flowEditorConfig = getFlowEditorConfig(flowType);
  const win = window as any;
  if (win.isMobile && win.isMobile()) {
    flowEditorConfig.mutable = false;
  }

  const configProviderContext = { config: flowEditorConfig };
  return configProviderContext;
};

export const setMock = (implementation?: (...args: any[]) => any): Query<jest.Mock> =>
  set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  flowType: FlowTypes = FlowTypes.MESSAGING,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = getConfigProviderContext(flowType)
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<ConfigProviderContext | Partial<ConfigProviderContext>> = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: Array<JSX.Element | React.ComponentClass | React.SFC> = []
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(baseDuxState, duxStateOverrides) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes
    }
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance()
  };
};

export const composeSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
  jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: Array<jest.SpyInstance<any>>) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec'
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  flowType: FlowTypes = FlowTypes.MESSAGING,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = getConfigProviderContext(flowType)
) => ({
  setup: composeSetup<P>(Component, baseProps, flowType, baseDuxState, baseContext),
  spyOn: composeSpy(Component)
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes
      }
    }
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(object: T, property: K, value: T[K]) => {
  Object.defineProperty(object, property, { get: () => value });
};

interface StoreOverrides {
  languageCode?: string;
  isTranslating?: boolean;
  results?: InfoResult[];
}

export const setupStore = ({
  languageCode = 'eng',
  isTranslating = false,
  results = []
}: StoreOverrides = {}) => {
  const languageNames: { [key: string]: string } = { eng: 'English', spa: 'Spanish' };
  const store = loadStore();
  mock(store, 'getState', () => {
    const state: TembaAppState = {
      isTranslating,
      languageCode,
      languageNames,
      workspace: {
        uuid: createUUID(),
        name: 'Test Workspace',
        languages: ['eng', 'spa'],
        timezone: 'UTC',
        date_style: 'MM/DD/YYYY',
        country: 'US',
        anon: false
      },
      flowDefinition: null,
      flowInfo: null,
      
      // Editor state from Redux
      editorState: null,
      
      // Flow context from Redux  
      flowNodes: {},
      flowMetadata: null,
      assetStore: null,
      flowIssues: {},
      
      // Node editor state from Redux
      nodeEditorSettings: null,
      nodeEditorTypeConfig: null,
      userAddingAction: false,
      
      setFlowInfo: (info: FlowInfo) => {},
      setFlowContents: (contents: FlowContents) => {},
      setLanguageCode: (code: string) => {},
      getLanguage: () => ({ name: languageNames[languageCode], code: languageCode }),
      getFlowResults: () => results,
      getResultByKey: (key: any) => {
        const result = results.find(r => r.key === key);
        return result || null;
      },
      updateCanvasPositions: (positions: any) => {},
      removeNodes: (uuids: string[]) => {},
      
      // Editor state management
      updateEditorState: (changes: any) => {},
      
      // Flow context management
      updateFlowDefinition: (definition: any) => {},
      updateFlowNodes: (nodes: any) => {},
      updateAssetStore: (assetStore: any) => {},
      updateFlowIssues: (issues: any) => {},
      updateFlowMetadata: (metadata: any) => {},
      
      // Node editor management
      updateNodeEditorSettings: (settings: any) => {},
      updateNodeEditorTypeConfig: (typeConfig: any) => {},
      updateUserAddingAction: (userAddingAction: boolean) => {}
    };
    return state;
  });

  mock(store, 'getApp', () => {
    const app: App = {
      subscribe: (callback: (state: any, prevState: any) => void) => {
        // Mock implementation of subscribe
        return () => {
          // Mock unsubscribe function
        };
      }
    };
    return app;
  });

  mock(store, 'resolveUsers', async (objects: any[], fields: string[]) => {
    // mock implementation to add UUIDs to users that don't have them
    objects.forEach(obj => {
      fields.forEach(field => {
        if (obj[field] && obj[field].email && !obj[field].uuid) {
          obj[field].uuid = `mock-uuid-${obj[field].email.replace('@', '-at-')}`;
        }
      });
    });
  });
};
