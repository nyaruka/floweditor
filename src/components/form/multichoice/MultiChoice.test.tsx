import { composeComponentTestUtils, mock } from 'testUtils';
import * as utils from 'utils';

import MultiChoiceInput, { MultiChoiceInputProps } from './MultiChoice';

const { setup } = composeComponentTestUtils<MultiChoiceInputProps>(MultiChoiceInput, {
  name: 'Multi Choice',
  items: { value: ['one', 'two', 'three'] },
  maxItems: 10
});

mock(utils, 'createUUID', utils.seededUUIDs());

describe(MultiChoiceInput.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });
});
