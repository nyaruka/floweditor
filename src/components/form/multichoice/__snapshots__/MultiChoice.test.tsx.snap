// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiChoiceInput should render 1`] = `
<Fragment>
  <p />
  <TembaSelect
    expressions={true}
    maxitems={10}
    multi={true}
    name="Multi Choice"
    onChange={[Function]}
    placeholder="Multi Choice"
    searchable={true}
    tags={true}
    value={
      Array [
        Object {
          "name": "one",
          "value": "one",
        },
        Object {
          "name": "two",
          "value": "two",
        },
        Object {
          "name": "three",
          "value": "three",
        },
      ]
    }
  />
</Fragment>
`;
